File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        if (n <= 2) {
            return 0;
        }
        int maxLeft = height[0];      
        int maxRight = height[n - 1]; 
        int trappedWater = 0;
        int left = 1;
        int right = n - 2;

        while (left <= right) {
            if (maxLeft < maxRight) {
                if (height[left] >= maxLeft) {
                    maxLeft = height[left];
                } else {
                    trappedWater += maxLeft - height[left];
                }
                left++;
            } else {
                if (height[right] > maxRight) {
                    maxRight = height[right];
                } else {
                    trappedWater += maxRight - height[right];
                }
                right--;
            }
        }
        
        return trappedWater;
    }}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if n <= 2:
            return 0
        max_left = height[0]      
        max_right = height[n - 1] 
        trapped_water = 0
        left = 1
        right = n - 2
        while left <= right:
            if max_left < max_right:
                if height[left] >= max_left:
                    max_left = height[left]
                else:
                    trapped_water += max_left - height[left]
                left += 1
            else:
                if height[right] > max_right:
                    max_right = height[right]
                else:
                    trapped_water += max_right - height[right]
                right -= 1
        return trapped_water